Cypress: End-to-end and Integration tests, framework

ccjoel opened the pull request:
Needs some cleanup, but can get help on this after merging.

Follow-up PRs will come after, hopefully with help from others.
TODO: modify some hardcoded values for seeds from using localhost:8080 to using `Cypress.config('baseUrl')`, add ability to login with our stage auth. This is in order to run this tests against a deployed staged env instead of running E2E tests against local only.

brandomr commented on api/src/dmx.py:
```diff
@@ -0,0 +1,31 @@
+    logger.info(f'dmx examining {url} return_comments={return_comments}')
+    return dmx.examine(url, return_comments = return_comments)
+
+    #dmx_id = uuid.uuid4()
```
Let's clean up the commented stuff before merging.

GoogleSheets commented:
current version of `domainmodelexaminer` PyPi package is 0.1.1

https://pypi.org/project/domainmodelexaminer/0.1.1/

mattprintz commented:
@GoogleSheets If we skip fetching the comments, does the data fit well enough in to ElasticSearch? I've kind of changed my mind and think that, if we can fit it safely, then storing the values used to populate the fields would be good. This would allow us to later on compare how well DMX is performing vs the final product, and/or use to on future features that have yet to be determined.

mattprintz commented on api/src/dmx.py:
```diff
@@ -0,0 +1,31 @@
+logger = logging.getLogger(__name__)
+
+@router.post("/dmx/examine")
+def examine_model(url: str, return_comments: bool = True):
```
AFAIK, `async` will only ever do anything if the underlying code uses `async/await` internally. Our redis implementation does and I think that's most ,if not all, of where we use async.
For it to make a difference, I think the dmx library would require updates.

GoogleSheets replied:
If the endpoint returns json, it should probably not be `async` and just spin. If it indexes the data in elastic search, then `async` seems much better.

mattprintz commented on api/src/dmx.py:
```diff
@@ -0,0 +1,31 @@
+logger = logging.getLogger(__name__)
+
+@router.post("/dmx/examine")
+def examine_model(url: str, return_comments: bool = True):
```
For the Dojo endpoint, I think that the default value for `return_comments` should be False, since I can't really foresee us using information from the comments to make any decisions like we might with the rest of the information.

ccjoel committed:
Re-enabled and adapted a subset of the model registration,listing E2E tests for them to pass. Added libs for seeding, mocking, faking, cleaning, as well as utils on cypress to setup, run, clean E2E tests. Needs some cleanup and more tests pending.

ccjoel committed:
Added new ignore pattern for e2e (with backend) tests so that they dont run on
CI (github actions) for now, since they'll fail.

ccjoel committed:
Have a draft test that seeds a dataset, then runs elwood/publish and asserts
against its final state. final assertion pending.

ccjoel committed:
acled redacted file ws missing

ccjoel committed:
Added crazy high timeouts while optimizing for the e2e tests to run on slower
systems (such as soc macos+docker).

ccjoel committed:
Added better cy task nodejs scripts debugging, reusing code for seeding/setup.
Tests passing.

ccjoel committed:
cleaning up some more dataset/model tests

ccjoel committed:
Moved shared cypress stub http calls code to support/helpers. Added timeout when
shutting down terminal workers for model registration.

ccjoel committed:
Re-enabled/configured cypress component tests

ccjoel committed:
templater test

ccjoel committed:
renamed full screen templater file to e2e so that we may skip when necessary

ccjoel committed:
Model annotate from terminal tests.

ccjoel committed:
creating a new helper plugin. well see if it works

ccjoel committed:
More tests- directives panel.

ccjoel committed:
Model directive editing tests.

ccjoel committed:
All tests passing, with proper seeding and cleanup.

ccjoel committed:
Added some waits while we discuss how to handle some weird app failure scenarios.

ccjoel committed:
Oops, have commited huge csv file which we won't use.

ccjoel committed:
Tests passing, will add async helpers outside of tests.

ccjoel committed:
Partial updates.. this refactor is tricky.

ccjoel committed:
Try this on macbook.

ccjoel committed:
some cleanup for wait for elwood fn

ccjoel committed:
More cleanup,timing on macos

ccjoel committed:
Moved helper to helpers file.

ccjoel committed:
Merge branch 'e2e-merge-async-strategies-final' of github.com:jataware/dojo into e2e-merge-async-strategies-final

ccjoel committed:
Fixed await for run elwood on dataset publish e2e tests.

ccjoel committed:
Test Model Tests updated.

ccjoel commented:
The exludeSpecPattern didn't kick in. Maybe because this branch doesn't have latest main merged in. Will retry soon.

ccjoel committed:
Updated github hook exclude spec pattern to ignore e2e tests.

ccjoel committed:
Test Model registered and published correctly.

ccjoel committed:
Upgrades to test_model E2E tests.

ccjoel committed:
Forcing to NOT run on github actions using 2 different methods.

ccjoel committed:
Added stub-mock endpoint for an integration test on datset update flow. Added
some debugging info to display full CYPRESS config when running on CI. Added
some logging to explain dev what the test is doing while running on Cypress IDE.

ccjoel commented on .github/workflows/int_tests_cypress.yml:
```diff
@@ -36,7 +36,7 @@ jobs:
         with:
           working-directory: ./ui
-          config: video=false,videoUploadOnPasses=false,keystrokeDelay=1,retries=2
+          config: '{"e2e": { "excludeSpecPattern": ["*.hot-update.js", "**/*.e2e.cy.js"] }, "video": false, "keystrokeDelay": 1, "retries": 2,  "component": {"excludeSpecPattern": ["*.hot-update.js", "**/*.e2e.cy.js"]}}'
```
Couldn't get it to work as before (inline config).. had to repalce with JSON. We'll clean up if we find an alternative, but this was blocking this PR.

ccjoel commented on ui/.eslintrc.js:
```diff
@@ -1,6 +1,7 @@
   "parser": "@babel/eslint-parser",
-  "extends": ["airbnb", "plugin:react-hooks/recommended", "plugin:storybook/recommended"],
+  "extends": ["airbnb", "plugin:react-hooks/recommended",
+              "plugin:storybook/recommended", "plugin:cypress/recommended"],
```
We still don't have a cypress lint task. Follow-up work to add it and address problems!

ccjoel commented on ui/client/components/templater/TextAnnotater.js:
```diff
@@ -87,6 +87,7 @@ const Mark = ({
   const displayInnerContent = () => (
     <span
       className={savedHover ? classes.darkerMark : classes.mark}
+      data-testid="annotation-inner-contents"
```
This is testid and not test, as it's not a custom `cy.get(''[data-test=...)` and instead it is a built-in plugin: `cy.findByTestId(...)`. We'll consolidate in follow-up work as needed.

ccjoel commented on ui/client/datasets/RunJobs.js:
```diff
@@ -122,7 +122,7 @@ const RunJobs = withStyles(({ spacing }) => ({
     } else {
       // No result, wait for an update
-      timeoutHandle = setTimeout(updateJobData, 500);
+      timeoutHandle = setTimeout(updateJobData, 1500);
```
This makes the user wait an additional second once a job has finished, but helps both the tests and the deployment instance not go crazy memory-wise.

ccjoel commented on ui/cypress/e2e/README.md:
```diff
@@ -0,0 +1,31 @@
+Depends on host:
+cy.onlyOn('localhost')
+
+onlyOn('mac', () => {
```
Argh, was looking for this earlier in the docs, couldn't find it. I'll add a proper helper README as follow-up work. Help needed.

ccjoel commented on ui/cypress/e2e/test_model.e2e.cy.js:
```diff
@@ -0,0 +1,243 @@
+        cy.get('@TerminalInput')
+          .type(`sudo apt update && sudo apt install -y python3-pip && git clone https://github.com/jataware/test-model.git && cd test-model{enter}`, {force: true});
+
+        cy.wait(25000); // 60000 for slower computers
```
This has a bunch of overlap with `model.e2e.cy.js` (not missing `test_` prefix on file). This particular file tests registering our jataware/test-model, which means we have to type on the virtual xterm terminal and wait for long commands to finish (sudo apt update; sudo apt install ...; pip install ...). I don't know of a proper way to verify once a command in this ssh-virtual terminal is completed, so we have arbitrary waits here. You'll have to change it in order to work on your system. This test is a bit overkill, but really auto-tests registering the test model. I still recommend running all the model.e2e.cy.js tests (and component tests!) as it really tests all aspects of model registration piecemeal.

ccjoel replied:
Help needed in order to check when the last running command has completed, without waits?

Maybe:

```
cy.get('@TerminalInput')
// select ^ this element, but with a not:disabled check, and add a very long timeout for it? Not great, but maybe better.
```

ccjoel commented on ui/cypress/plugins/index.js:
```diff
@@ -27,5 +150,78 @@ module.exports = (on, config) => {
 
+  on('task', {
+
+    'debug': (...args) => {debug(...args); return true;},
```
This is to get access to this really nice node-console debug on tests and on when running on github!

Use like:

```
cy.task('debug', myVar);
```

Where `myVar` can be any param allowed to `console.log`.

Locally, this will be printed on the terminal log for cypress, like:

![Screenshot from 2023-08-18 11-48-04](https://github.com/jataware/dojo/assets/1967061/36b637f3-da85-4765-92e9-c6303e3c24a0)

and on github action jobs:

![Screenshot from 2023-08-18 11-48-52](https://github.com/jataware/dojo/assets/1967061/12d2469a-fe4e-4c89-8cf6-a3122579dcd8)

ccjoel commented on ui/cypress/support/commands.js:
```diff
@@ -80,3 +83,35 @@ Cypress.Commands.add('setSelection', { prevSubject: true }, (subject, query, end
+
+// let sharedStatus = 0;
+
+// Cypress.Commands.add('spyPoll', {}, (method, url, property, expectedVal, nth=0) => {
```
Previous attemp to loop and wait until, for example, an elwood job completed.

Freakin nightmare. I opted to use a separate helpers.js function with async/await and do `cy.request()`s, then use `cy.wrap()` to wait for this promise in tests.

Help needed to consider replacing the `cy.request()` to `axios` and internal helpers.js `cy.wait()` to a `sleep()` function that uses promises, maybe.

ccjoel commented on ui/cypress/plugins/index.js:
```diff
@@ -27,5 +150,78 @@ module.exports = (on, config) => {
+
+    // TODO this is unused for now as we can use API
+    // for simple cases, but for complex scenarios it will be useful
+    'seed': ({type, id, name}) => {
```
My vision was to use `cy.task('seed'...)` everywhere on E2E tests so as to abstract away this chore, but didn't work as I wanted and is kinda a nice-to-have. Help needed. That is, change the test calls to seed from `cy.request(datasetUrl, data)... manual seed` to using `cy.task('seed', {type: 'dataset', ...)`

ccjoel commented on ui/cypress/seeds/model_api_data.js:
```diff
@@ -0,0 +1,250 @@
+    },
+    // "next_version": "string", // would mean the current version is deprecated in favor of next version
+    // "prev_version": "string",
+    "is_published": true,
```
Mocking this as published for the model list page, but then caught me by surprise (badly!) when reusing for the actual register tests (as the model was already "published" with this flag.).

TODO: create a second model seed mock fn based on this, maybe called `genUnregisteredModel`, and use that in the appropriate tests. Help needed.
